core/os/os2/process_linux.odin

```diff
@@ -71,25 +307,303 @@ _Sys_Process_Attributes :: struct {}
+	//       3. allocate some stack space
+	//       4. vfork (waits for child exit or execve); In child:
+	//           a. set child signal handlers
+	//           b. set up any necessary pipes
+	//           c. execve
+	//       5. restore signal handlers
+	//
+	pid: linux.Pid
+	if pid, errno = linux.fork(); errno != .NONE {
+		return process, _get_platform_error(errno)
+	}
+
+	STDIN  :: linux.Fd(0)
+	STDOUT :: linux.Fd(1)
+	STDERR :: linux.Fd(2)
+
+	if pid == 0 {
+		// in child process now
+		if desc.stdin != nil {
+			fd := linux.Fd(fd(desc.stdin))
+			if _, errno = linux.dup2(fd, STDIN); errno != .NONE {
+				linux.exit(1)
+			}
+		}
+		if desc.stdout != nil {
+			fd := linux.Fd(fd(desc.stdout))
+			if _, errno = linux.dup2(fd, STDOUT); errno != .NONE {
+				linux.exit(1)
+			}
+		}
+		if desc.stderr != nil {
+			fd := linux.Fd(fd(desc.stderr))
+			if _, errno = linux.dup2(fd, STDERR); errno != .NONE {
+				linux.exit(1)
+			}
+		}
+
+		if errno = linux.execveat(dir_fd, executable_path, &cargs[0], env); errno != .NONE {
+			print_error(stderr, _get_platform_error(errno), string(executable_path))
+			panic("execve failed to replace process")
+		}
+		unreachable()
+	}
+
+	process.pid = int(pid)
+	process.handle = PIDFD_UNASSIGNED
+	return
+}
+
+_process_state_update_times :: proc(p: Process, state: ^Process_State) -> (err: Error) {
+	TEMP_ALLOCATOR_GUARD()
+
+	stat_path_buf: [32]u8
+	_ = fmt.bprintf(stat_path_buf[:], "/proc/%d/stat", p.pid)
+	stat_buf: []u8
+	stat_buf, err = _read_entire_pseudo_file(cstring(&stat_path_buf[0]), temp_allocator())
+	if err != nil {
+		return
+	}
+
+	// ')' will be the end of the executable name (item 2)
+	idx := strings.last_index_byte(string(stat_buf), ')')
+	stats := string(stat_buf[idx + 2:])
+
+	// utime and stime are the 14 and 15th items, respectively, and we are
+	// currently on item 3. Skip 11 items here.
+	for _ in 0..<11 {
+		stats = stats[strings.index_byte(stats, ' ') + 1:]
+	}
+
+	idx = strings.index_byte(stats, ' ')
+	utime_str := stats[:idx]
+
+	stats = stats[idx + 1:]
+	stime_str := stats[:strings.index_byte(stats, ' ')]
+
+	utime, _ := strconv.parse_int(utime_str, 10)
+	stime, _ := strconv.parse_int(stime_str, 10)
+
+	// NOTE: Assuming HZ of 100, 1 jiffy == 10 ms
+	state.user_time = time.Duration(utime) * 10 * time.Millisecond
+	state.system_time = time.Duration(stime) * 10 * time.Millisecond
+
 	return
 }
 
 @(private="package")
 _process_wait :: proc(process: Process, timeout: time.Duration) -> (process_state: Process_State, err: Error) {
+	process_state.pid = process.pid
+
+	options: linux.Wait_Options
+	big_if: if timeout == 0 {
+		options += {.WNOHANG}
+	} else if timeout > 0 {
+		ts: linux.Time_Spec = {
+			time_sec  = uint(timeout / time.Second),
+			time_nsec = uint(timeout % time.Second),
+		}
+
+		// pidfd_open is fairly new, so don't error out on ENOSYS
+		pid_fd: linux.Pid_FD
+		errno: linux.Errno
+		if _has_pidfd_open() {
+			if process.handle == PIDFD_UNASSIGNED {
```

Comment:
Shouldn't `process_start` open the pidfd? I feel like once you get the `Process` you automatically should get the pidfd if its available. That's sort of the point of the process handle, as long as you have it, it should refer to the same process. This can probably also simplify the caching code, if not completely remove it.

Reply:
Agreed. I think I was worried about waiting on a process that isn't my child, but that should never happen. This removes the need for caching.